### PR TITLE
Implement the power operation #3

### LINKED ISSUE #3: Task: Implement the power operation
It is required to implement the power operation.

# Hints

Replace the following two instructions in the `power` function with your implementation:

```
  la $a0, unimplemented_message
  jal printString
```

Use the available utilities to read and write the numbers. Use the subtract function as a reference.

### PR DESCRIPTION
**Pull request by MohamedAbdallah17**
Resolves #3.

### CODE REVIEW

File: mainMARS.s [hunk 1]
```diff
@@ -217,12 +217,57 @@ max:
...
+  move $t2, REG_READ_INTEGER_RET
+
+  # Check if the exponent is zero.
+  lwc1 $f3, zero_float
+  c.eq.s $f2, $f3
+  bc1f zero_exponent
+    la REG_PRINT_STRING_ARG,zero_exponent_message
+    jal printString
+    pop_ra_and_return
+  zero_exponent:
```
Comment (OmarEmaraDev) [inline mainMARS.s]: You probably don't need a special branch for this. See next comment.

File: mainMARS.s [hunk 2]
```diff
@@ -217,12 +217,57 @@ max:
...
+
+  # Power
+  add $t0, $zero, $zero 	#intialize $t0=0 , $t0 is used to record how many times we do multiplication operation  
+  add $t3, $t3, 1		#set intial value of $t3 = 1
+  power_loop_start:
+    beq $t0, $t2, power_loop_end 
+    mul $t3, $t3, $t1		#multiplication of $t1 and $t4 into $t4  
+    addi $t0,$t0,1		#update the value of $t0
+    j power_loop_start
+  power_loop_end:
```
Comment (OmarEmaraDev) [inline mainMARS.s]: You can start from the value of `b` and subtract one every loop, that was, if `b` is zero, the loop will not be executed and the output will be 1. This is similar to how we do the factorial above.

File: mainMARS.s [hunk 3]
```diff
@@ -217,12 +217,57 @@ max:
...
-  push_ra
-  la REG_PRINT_STRING_ARG, unimplemented_message
+power:  
+  # Print the message for a.
+  la REG_PRINT_STRING_ARG, power_a_message
+  jal printString
+
+  # Read a.
+  jal readInteger
+  move $t1, REG_READ_INTEGER_RET
```
Comment (OmarEmaraDev) [inline mainMARS.s]: Don't store values in temporary registers if you will be calling functions after. Those registers are not callee-saved, meaning their values may change across function calls. Use saved registers instead.

File: mainMARS.s [hunk 4]
```diff
@@ -217,12 +217,57 @@ max:
...
+
+  # Print the message for b.
+  la REG_PRINT_STRING_ARG, power_b_message
+  jal printString
+
+  # Read b.
+  jal readInteger
+  move $t2, REG_READ_INTEGER_RET
```
Comment (OmarEmaraDev) [inline mainMARS.s]: Same as above. Also, this value maybe negative, which is a case you don't consider.

File: mainMARS.s [hunk 5]
```diff
@@ -217,12 +217,57 @@ max:
...
+    jal printString
+    pop_ra_and_return
+  zero_exponent:
+
+  # Print the result message.
+  la REG_PRINT_STRING_ARG, result_message
   jal printString
+
+  # Power
+  add $t0, $zero, $zero 	#intialize $t0=0 , $t0 is used to record how many times we do multiplication operation
```
Comment (OmarEmaraDev) [inline mainMARS.s]: This is equivalent to `move $t0, $zero`.
I would prefer if the comments where in their separate lines, not after the lines, like in the rest of the code.

**Review by OmarEmaraDev — CHANGES_REQUESTED** [review]

File: mainMARS.s (hunk 3 above)
Comment (MohamedAbdallah17) [reply mainMARS.s]: You mean that i would use **$s0 - $s7**

Comment (OmarEmaraDev) [reply mainMARS.s]: @MohamedAbdallah17 Yes.

File: mainMARS.s (hunk 2 above)
Comment (MohamedAbdallah17) [reply mainMARS.s:264]: when we say `mul $s3 ,$s1, s1` 
let's say we had a value saved in `$s3` ...is the result of multiplying `$s1` by `$s1` loaded into `$s3` neglecting the old value we had in `$s3` or adding it to old value ?.

Comment (MohamedAbdallah17) [reply mainMARS.s]: `# Power`
`  li $s3, $s1`
`  Power_loop_start:`
` li $t1, $zero`
 ` beqz $s2, Power_loop_end`
  `  mul $t1, $s3, $s1`
  `  add $s3, $s3, $t1`
  `  sub $s2, $s2, 1`
  `  b  Power_loop_start`
`  Power_loop_end:`
is this what you meant or not ? `$s1` is `a` and `$s2` is `b` and `$s3` is the result that would be printed after that 
How could i consider the zero case more efficiently not like the way i implemented it before ,, and i'll finish the case of **if b is negative value** and provide it for discussing

Comment (OmarEmaraDev) [reply mainMARS.s]: @MohamedAbdallah17 You can incorporate the zero case by initializing `$s3` to 1 instead. The following C code demonstrates the solution:
```c
int power(int a, int b) {
    int result = 1;
    while (1) {
        if (b == 0) {break;}
        result *= a;
        b--;
    }
    return result;
}
```
A for loop can be used here, but I used a while loop for easier translation to assembly.

File: mainMARS.s [hunk 6]
```diff
@@ -217,12 +217,64 @@ max:
...
 
 # Power operation.
+.data 
+power_a_message: .asciiz "Enter the a in (a ^ b):\n"
+power_b_message: .asciiz "Enter the b in (a ^ b):\n"
+zero_exponent_message: .asciiz "The result equals 1 "
 
 .text
-power:
-  push_ra
```
Comment (OmarEmaraDev) [inline mainMARS.s:223]: `push_ra` is necessary here. Don't remove it.

File: mainMARS.s [hunk 7]
```diff
@@ -217,12 +217,64 @@ max:
...
+
+  # special case if b equals a negative value.3
+  move $s4, $s1
+  Special_power_loop_start:
+  bltz $s2, Special_power_loop_end
+    mul $t1, $s3, $s1
+    add $s4, $s4, $t1
+    add $s2, $s2, 1
+    b Special_power_loop_start
+  Special_power_loop_end:
```
Comment (OmarEmaraDev) [inline mainMARS.s]: Negative input handing is much simpler than this.
1. If b is negative and a is zero, print an error message and return.
2. Store the sign bit of b.
3. Take the absolute value of the b.
4. Compute the power.
5. If the sign bit is set, compute the reciprocal of the power.

File: mainMARS.s [hunk 8]
```diff
@@ -217,12 +217,64 @@ max:
...
+  
+  # Print the result message.
+  li $t1, 1
+  div $s5, $t1, $s4
```
Comment (OmarEmaraDev) [inline mainMARS.s:256]: `div` can't be used here. You need to move the value to Co-Processor 1, compute the reciprocal there using the `RECIP.S` instruction and print the float.

File: mainMARS.s [hunk 9]
```diff
@@ -217,12 +217,64 @@ max:
...
+  Power_loop_start:
+  move $t2, $zero
+  beqz $s2, Power_loop_end
+    mul $t2, $s3, $s1
+    add $s3, $s3, $t2
+    sub $s2, $s2, 1
+    b  Power_loop_start
+  Power_loop_end:
+  # Print the result message.
+  la REG_PRINT_STRING_ARG, zero_exponent_message
```
Comment (OmarEmaraDev) [inline mainMARS.s]: Why are you printing this here? It seems like you are printing multiple results here.

**Review by OmarEmaraDev — CHANGES_REQUESTED** [review]

File: mainMARS.s (hunk 6 above)
Comment (MohamedAbdallah17) [reply mainMARS.s:223]: sorry I don't mean to remove it

**Review by OmarEmaraDev — APPROVED** [review]
Negative input handing will be refactored in a follow up commit.

Thanks!

**Pull request merged**